src/core.c/Rakudo/Internals.pm6

```diff
@@ -1384,15 +1384,15 @@ my class Rakudo::Internals {
             nqp::while(
               nqp::chars(my str $entry = self!next),
               nqp::stmts(
+                (next if $entry eq '.' or $entry eq '..'),
                 (my str $path = nqp::concat($!abspath,$entry)),
                 (try
                   nqp::if(
-                    $!file.ACCEPTS($entry) &&
-                      nqp::stat($path,nqp::const::STAT_ISREG),
+                      (my $s := nqp::dispatch('boot-syscall', 'file-stat', nqp::decont_s($path), 0)) && nqp::dispatch('boot-syscall', 'stat-is-reg', $s) && $!file.ACCEPTS($entry),
```

Comment:
It might be more clear putting the `my $s := nqp::dispatch('boot-syscall', 'file-stat', nqp::decont_s($path), 0))` as its own line in the `nqp::stmts`